fix: detect completed tasks even if console connects after

seanmonstar opened the pull request:
The "completion" of a task would result in that being sent to all watchers in a task update. However, after a flush, that status wasn't stored in a way that new watchers could detect. This changes it so that a task is determined "completed" when it includes a "total" duration, instead of being in a separate part of the `TasksUpdate` proto message.

MSleepyPanda commented:
@hawkw most relevant is `subscriber/*` and the `proto/tracing.proto` definition

I'm not sure about the two thread design in the subscriber. I think it makes sense, but at the same time it feels like a hack. Feedback would be appreciated!

MSleepyPanda commented:
In case people want to try it out: run the example from `example/` via `cargo run` and start the console via `cargo run` in `console/` :)

hawkw commented on console/src/storage/store.rs:
```diff
@@ -0,0 +1,143 @@
+        let span = self.id_map[&record
+            .span
+            .as_ref()
+            .expect("BUG: No id set on follows.span")
```
Nit: I think you meant to say 
```suggestion
            .expect("BUG: No id set on record.span")
```

hawkw commented on subscriber/src/subscriber.rs:
```diff
@@ -0,0 +1,144 @@
+        id.clone()
+    }
+    fn drop_span(&self, ref id: span::Id) {
+        let mut registry = self.registry.lock().unwrap();
```
Note that [because `drop_span` can be called in `drop` impls](https://docs.rs/tracing-core/0.1.2/tracing_core/subscriber/trait.Subscriber.html#method.try_close), it should probably not contain any code that may panic; if it does, it will cause a double panic, resulting in the program being terminated with a SIGILL rather than displaying a panic message and backtrace. We should probably return early rather than panicking in this function.

See also jonhoo/tracing-timing#7 for a similar discussion.

MSleepyPanda replied:
We could also move to `parking_lot`, their Mutex'es have no notion of thread poisoning.

MSleepyPanda replied:
Actually i have encountered this double panic behaviour in an earlier stage, when i didn't account for connection resets! Very much in favor of solving this cleanly :)

hawkw replied:
> We could also move to `parking_lot`, their Mutex'es have no notion of thread poisoning.

I've also often observed better performance from `parking_lot` compared with the standard library's `Mutex` and `RwLock`, but that's a problem for later when we start thinking more about performance.

With that said, I think lock poisoning is, in general, a _good_ thing; since panics almost always represent a fatal, unrecoverable error, we _want_ to propagate that across threads in most cases. It's just that for functions called in `Drop` impls, in particular, the thread is _already_ panicking (possibly due to another function that unwrapped the poisoned lock).

FWIW, I wrote a little macro to reduce some of the boilerplate associated with handling a lock result when we might already be panicking; feel free to borrow it if you find it useful: https://github.com/tokio-rs/tracing/blob/5a6bd644f26ff62129e849094f297d1a3d33003c/tracing-subscriber/src/lib.rs#L4-L15

hawkw commented on subscriber/src/subscriber.rs:
```diff
@@ -0,0 +1,144 @@
+        }
+    }
+    fn clone_span(&self, id: &span::Id) -> span::Id {
+        self.registry.lock().unwrap().spans[SpanId::new(id.into_u64()).as_index()].refcount += 1;
```
Style nit, take it or leave it: this huge expression is, IMO, a bit unwieldy & hard to read; I would probably have written something like
```rust
let index = SpanId::new(id.into_u64()).as_index();
self.registry.lock().unwrap().spans[index].refcount += 1;
```
or similar.

(as a side benefit, the lock might be held for slightly less time)

hawkw commented on subscriber/src/lib.rs:
```diff
@@ -0,0 +1,99 @@
+
+#[derive(Debug)]
+pub struct Span {
+    refcount: usize,
```
If we represented a `Span`s ref count as an `AtomicUsize`, we'd be able to increment and decrement them with an immutable reference. We could then use a read-write lock for the registry, rather than a mutex, and only acquire a read lock to update reference counts. This would reduce lock contention pretty significantly; as ref-counting operations could take place concurrently.

hawkw commented on subscriber/src/subscriber.rs:
```diff
@@ -0,0 +1,144 @@
+
+pub struct ConsoleForwarder {
+    pub(crate) tx: Sender<Variant>,
+    pub(crate) registry: Arc<Mutex<crate::Registry>>,
```
We should look into using a `RwLock` rather than a `Mutex` here, since I think there are enough potential concurrent reads that this could make a significant performance improvement (especially if we used atomic integers for reference counts; as I mentioned above).

It would be okay to come back and make this kind of performance optimization later, since this is currently a prototype.

hawkw commented on subscriber/src/server.rs:
```diff
@@ -0,0 +1,139 @@
+
+#[derive(Clone)]
+/// A factory for ConsoleForwarder
+pub struct BackgroundThreadHandle {
```
Since `tracing` will often be used from asynchronous applications using futures, what do you think about implementing the server as a background _task_, rather than as a background _thread_? The background work could be a future spawned on an executor, and we could use futures-aware channels rather than `std` channels to communicate with it. In cases where users are not using futures and do want to have a dedicated background thread, we could wrap that future with a dedicated executor that runs in its own thread.

This approach is okay for now as a prototype, but we should consider changing this eventually.

MSleepyPanda replied:
This is what i meant with, "feels like a hack"

I very much agree that this is the right aproach, the honset answer is that i'm not fluent enough with futures/tokio to pull this off in one go, i will change this in a follow up, but i will need some one-on-one time with the tokio docs :D

hawkw replied:
> I very much agree that this is the right aproach, the honset answer is that i'm not fluent enough with futures/tokio to pull this off in one go, i will change this in a follow up, but i will need some one-on-one time with the tokio docs :D

I'm also happy to provide guidance on this when you get to that point!

hawkw commented on subscriber/src/lib.rs:
````diff
@@ -0,0 +1,99 @@
+//!  
+//! # Thread overview:
+//!
+//! ```schematic,ignore
````
nice diagram! :)

hawkw commented on subscriber/src/server.rs:
```diff
@@ -0,0 +1,139 @@
+#[derive(Default)]
+pub(crate) struct Registry {
+    pub spans: Vec<Span>,
+    pub reusable: Vec<SpanId>,
```
A potential future optimization to think about (not necessary for now) is to store the free list of reusable IDs in the same `Vec` as active spans. We could use an enum like 

```rust
enum SpanState {
    Active(Span),
    Free { next_id: SpanId },
}
```
and then track the most recently freed ID as a single `SpanId` field on the registry struct. When a span in the `spans` vec is closed, we replace it with the `SpanState::Free` variant with the `next_id` field set to the previous value of `Registry.next_id`, and set `Registry.next_id` to the closed span's ID. This would probably use space more efficiently. The `tracing-fmt` crate [uses a similar strategy](https://github.com/tokio-rs/tracing/blob/master/tracing-fmt/src/span.rs).

MSleepyPanda replied:
Honsetly, i'm impressed!

If we change `InternalId`s usize to `NonZeroUsize` we might also enable niche optimizations, making the enum tag basically free, while saving `3 (Vec) - 1(.next_id) = 2` words

MSleepyPanda commented on console/src/storage/store.rs:
```diff
@@ -0,0 +1,143 @@
+impl Store {
+    fn new_span(&mut self, span: NewSpan) {
+        // Update id mapping for span, see `Store` documentation
+        self.id_map.insert(
```
The subscriber won't guarantee that - the console itself upholds this invariant, exactly with this insert. This means when a span id is reused by the subscriber, this insert will return `Some(span)`.
I should stress on that in the documentation.

On the other hand, since we don't transmit `DropSpan`s, this means the mapping will not eagerly remove a now unused id, but that's fine, since the subscriber will also not reclaim that memory, but store that the index/id for reusage.

hawkw commented:
> In case people want to try it out: run the example from `example/` via `cargo run` and start the console via `cargo run` in `console/` :)

FYI: if you put the example in an `examples/` dir in an existing crate (it should probably be `subscriber/examples` IMO), it can be run from the workspace root with `cargo run --example`, and the console can be run from the root with `cargo run --bin console`. 

Here's a [blog post](http://xion.io/post/code/rust-examples.html). Eventually it might be nice to document how to run the console & examples in a README...

hawkw commented on console/src/storage/store.rs:
```diff
@@ -0,0 +1,143 @@
+
+    updated: bool,
+    id_counter: usize,
+    id_map: HashMap<u64, InternalId>,
```
It looks like both `id_counter` and `id_map` are only written _or_ read from in methods on `StoreHandle`; and never read from the UI thread. As far as I can tell, the UI thread will never need to touch them, as the thread that receives updates from the server handles the ID translation?

In that case, we should probably move them out of the shared state so that they don't need to be locked.

hawkw commented on console/src/storage/store.rs:
```diff
@@ -0,0 +1,143 @@
+
+    /// Locks and updates the underlying `Store`
+    pub fn handle(&self, variant: Variant) {
+        let mut store = self.0.lock().unwrap();
```
Looking at the various functions this method calls, it seems like they have different requirements about being able to read and write to different parts of the registry. For example, the `event` method is the only one that accesses the `store.events` field, while the `store.spans` field is written and read from in other methods. It seems to me that these should be locked independently, so that the UI thread can read events while we are writing to the spans, and so on.

It seems to me that it's probably a good idea to introduce more granular locking on the different parts of the registry. For example, the `Store::event` method doesn't need to lock the `store.spans` field. We could also consider using a read-write lock on the

hawkw commented on console/src/ui/events.rs:
```diff
@@ -114,14 +90,36 @@ impl EventList {
+                    Some(value::Value::Debug(d)) => write!(text, "{}", d.debug).unwrap(),
+                    None => {}
+                }
+                text.push_str(r#"")"#);
```
Looks like an extra quotation mark slipped in here --- when I run the console, all event fields render like this:
```
│DEBUG message(hello! I'm gonna shave a yak."), excitement(yay!")
│TRACE message(yak shaved successfully")
│TRACE yak(1"), shaved(true")
│TRACE yaks_shaved(1")
```

I think the solution is just to remove it
```suggestion
                text.push_str(r#")"#);
```

hawkw commented on proto/tracing.proto:
```diff
@@ -0,0 +1,113 @@
+
+message Record {
+  SpanId span = 1;
+  repeated Value values = 2;
```
What do you think about representing field-value pairs with a protobuf [map](https://developers.google.com/protocol-buffers/docs/proto3#maps) instead? We could then remove the `Field` from the `Value` message.
```suggestion
  map<Field, Value> values = 2;
```

MSleepyPanda replied:
This does not work, as protobuf maps may not have messages as fields.

hawkw replied:
Ah, really? Nevermind me then!

hawkw commented on proto/tracing.proto:
```diff
@@ -0,0 +1,113 @@
+
+message Event {
+  SpanId span = 1;
+  repeated Value values = 2;
```
Similarly, this could be a map.

hawkw commented on proto/tracing.proto:
```diff
@@ -0,0 +1,113 @@
+  }
+}
+
+message Attributes {
```
For what it's worth, I don't think it's strictly necessary for the messages to be _exactly_ the same structure as the corresponding `tracing` types --- we don't really need an `Attributes` message type, since it's only ever part of the `NewSpan` message type. The `Attributes` struct only exists in `tracing` for API design reasons, not because it represents a core concept that's central to the data model.

Instead, I might consider factoring out the attributes that are common between the `NewSpan` and `Event` messages into a new message, like
```protobuf
message Attributes {
  Metadata metadata = 1;
  bool is_root = 2;
  bool is_contextual = 3;
  SpanId parent = 4;
}
````

hawkw commented on proto/tracing.proto:
```diff
@@ -0,0 +1,113 @@
+  string file = 6;
+  LineNum line = 7;
+  bool is_event = 8;
+  bool is_span = 9;
```
I don't know if we really need to send the `is_event` and `is_span` fields on the wire; they are primarily used so that when a `Subscriber` is passed a bare `Metadata` ("bare" as in "separate from a Span or Event) in `enabled` and `register_callsite`, it knows what that metadata will correspond to. Here, a `Metadata` message is only ever part of a `NewSpan` or `Event` message, so we will always know what type of trace data it describes.

MSleepyPanda replied:
This was a mechanical translation, makes sense!

MSleepyPanda commented:
Same ci failure as #32

seanmonstar committed:
recognize completed tasks by stats updates

When a stats update includes the total duration of a task, we use that
to determin the task is completed.

seanmonstar committed:
remove completed tasks from proto message definition

hawkw commented on console/src/tasks.rs:
```diff
@@ -362,6 +355,14 @@ impl Task {
             .unwrap_or_else(|| self.total(since) - self.busy())
     }
+
+    fn updated(&mut self) {
```
very nitpicky, but, since this mutates the task in place, i would probably call it
```suggestion
    fn update(&mut self) {
```
i would probably expect something called `updated` to take a `Task` by value and return it.

hawkw replied:
also a nit, but it looks like this currently only handles marking tasks as completed. my guess is that the intention is for it to also eventually do other task state updates, but if that's not the case, it might be worth giving it a name that's specific to completion?

hawkw commented on proto/tasks.proto:
```diff
@@ -37,22 +37,11 @@ message TaskUpdate {
     // *is* included in this map, the corresponding value represents a complete
     // snapshot of that task's stats at in the current time window.
     map<uint64, Stats> stats_update = 3;
-    // The IDs of any tasks which have *completed* since the last update.
-    //
-    // If a task's ID is present in this list, that task has terminated. The
-    // client is free to delete any data it has stored for that task if it is no
-    // longer relevant. Task IDs *may* be reused when new tasks are spawned; if
-    // a new task is spawned for the same ID, it will appear in the `new_tasks`
-    // list in a future `TaskUpdate`.
-    //
-    // When a task ID is present in this list, the `stats_update` field of this
-    // `TaskUpdate` will contain the final statistics recorded for that task.
-    repeated common.SpanId completed = 4;
     // The system time when this update was recorded.
     //
     // This is the timestamp any durations in the included `Stats` were
     // calculated relative to.
-    google.protobuf.Timestamp now = 5;
```
just a note: when we stabilize the protobufs (e.g. when we publish a release of `console-api` on crates.io), we will want to stop making proto changes like this; instead, we will want to remove fields [by marking their field numbers as `reserved`](https://developers.google.com/protocol-buffers/docs/proto3#updating). for now, this is fine, since we aren't yet guaranteeing interoperability across git hashes.

hawkw reviewed (approved):
this looks good to me!

as a follow-up, we will probably want to make a change on the aggregator side to remove completed tasks once they have been sent in a task update with the `closed_at` time set, so that the hashmap of span data doesn't grow infinitely...

hawkw commented on console/src/tasks.rs:
```diff
@@ -126,14 +126,15 @@ impl State {
                 stats,
                 completed_for: 0,
             };
+            task.updated();
```
again, i would call this
```suggestion
            task.update();
```

hawkw commented on console/src/tasks.rs:
```diff
@@ -142,17 +143,9 @@ impl State {
-                tracing::warn!(?id, "tried to complete a task that didn't exist");
+                let mut t = task.borrow_mut();
+                t.stats = stats.into();
+                t.updated();
```
```suggestion
                t.update();
```

hawkw committed:
rename `updated` to `update`
